src/snowflake/snowpark/_internal/type_utils.py

```diff
@@ -699,6 +723,95 @@ def snow_type_to_dtype_str(snow_type: DataType) -> str:
     raise TypeError(f"invalid DataType {snow_type}")
 
 
+def retrieve_func_defaults_from_source(
+    file_path: str,
+    func_name: str,
+    class_name: Optional[str] = None,
+    _source: Optional[str] = None,
+) -> Optional[List[Optional[str]]]:
+    """
+    Retrieve default values assigned to optional arguments of a function from a
+    source file, or a source string (test only).
+
+    Returns list of str(default value) if the function is found, None otherwise.
+    """
+
+    def parse_default_value(
+        value: ast.expr, enquote_string: bool = False
+    ) -> Optional[str]:
+        if isinstance(value, (ast.Tuple, ast.List)):
+            return f"{[parse_default_value(e) for e in value.elts]}"
+        if isinstance(value, ast.Dict):
+            key_val_tuples = [
+                (parse_default_value(k), parse_default_value(v))
+                for k, v in zip(value.keys, value.values)
+            ]
+            return f"{dict(key_val_tuples)}"
+        if isinstance(value, ast.Attribute):
+            return f"{parse_default_value(value.value)}.{value.attr}"
+        if isinstance(value, ast.keyword):
+            return f"{value.arg}={parse_default_value(value.value)}"
+        if isinstance(value, ast.Constant):
+            if isinstance(value.value, str) and enquote_string:
+                return f"'{value.value}'"
+            if value.value is None:
+                return None
+            return f"{value.value}"
+        if isinstance(value, ast.Name):
+            return value.id
+        if isinstance(value, ast.Call):
+            parsed_args = ", ".join(
+                parse_default_value(arg, True) for arg in value.args
+            )
+            parsed_kwargs = ", ".join(
+                parse_default_value(kw, True) for kw in value.keywords
+            )
+            combined_parsed_input = (
+                f"{parsed_args}, {parsed_kwargs}"
+                if parsed_args and parsed_kwargs
+                else parsed_args or parsed_kwargs
+            )
+            return f"{parse_default_value(value.func)}({combined_parsed_input})"
+        raise TypeError(f"invalid default value: {value}")
```

Comment:
nit: is it possible to give a small code snippet of sample value in the comment in each if-condition so that it's easier to understand the map between the processed value and ast type

Reply:
sure. I'll add it.